go/database/mpt/proof.go

```diff
@@ -0,0 +1,197 @@
+package mpt
+
+import (
+	"bytes"
+	"errors"
+	"github.com/Fantom-foundation/Carmen/go/common"
+)
+
+//go:generate mockgen -source proof.go -destination proof_mocks.go -package witness
+
+// RlpNode is an RLP encoded MPT node.
+type RlpNode []byte
+
+// ProofElement is an entry of a witness proof.
+// It contains a hash of an MPT node and the node itself
+// encoded as an RLP byte slice.
+type ProofElement struct {
+	hash common.Hash
+	node RlpNode
+}
+
+// Equal returns true if the input proof element is equal to the current proof element.
+func (p ProofElement) Equal(other ProofElement) bool {
+	return p.hash == other.hash && bytes.Equal(p.node, other.node)
+}
+
+// Hash returns the hash of the proof element.
+func (p ProofElement) Hash() common.Hash {
+	return p.hash
+}
+
+// Rlp returns the RLP encoded node of the proof element.
+func (p ProofElement) Rlp() RlpNode {
+	return p.node
+}
+
+// proof is a chain of MPT nodes and their hashes that represent a witness proof.
+// It is used to verify the proof by checking that the nodes of the MPT are present in this proof.
+// The proof is created by iterating the MPT from the root down to the leaf node following
+// the address or a storage key.
+// Each node encountered on the path is hashed and stored in the proof.
+type proof []ProofElement
+
+// nodeVisitor is a visitor that visits MPT nodes and creates a witness proof.
+// It uses the hasher and node source to hash and encode the nodes.
+// It computes RLP and hash of nodes that compose the proof.
+type nodeVisitor struct {
```

Comment:
The name `nodeVisitor` is very generic and it is visible in the entire MPT package. Can you rename it to something like `proofExtractionVisitor`?

Reply:
done